tpm2/src/util/kmip_util.c

```diff
@@ -0,0 +1,584 @@
+                                      &key_request, &key_request_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to build the KMIP Get request.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  unsigned char *encrypted_request = NULL;
+  size_t encrypted_request_len = 0;
+
+  result = aes_gcm_encrypt(session_key, session_key_len,
+                           key_request, key_request_len,
+                           &encrypted_request, &encrypted_request_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the KMIP key request.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Sending request for a key with ID: %.*s", key_id_len,
+            key_id);
+  ssize_t write_result =
+    write(socket_fd, encrypted_request, encrypted_request_len);
+
+  if (write_result != encrypted_request_len)
+  {
+    kmyth_log(LOG_ERR, "Failed to fully send the key request.");
+    kmyth_log(LOG_ERR, "Expected to write %zd bytes, only wrote %zd bytes.",
+              encrypted_request_len, write_result);
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  // Read response from B; decrypt with S
+  unsigned char *encrypted_response = calloc(8192, sizeof(unsigned char));
+  size_t encrypted_response_len = 8192 * sizeof(unsigned char);
+
+  int read_result = read(socket_fd, encrypted_response, encrypted_response_len);
+
+  if (-1 == read_result)
+  {
+    kmyth_log(LOG_ERR, "Failed to read the key response.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  kmyth_log(LOG_INFO, "Received %zd bytes.", read_result);
+
+  unsigned char *response = NULL;
+  size_t response_len = 0;
+
+  result = aes_gcm_decrypt(session_key, session_key_len,
+                           encrypted_response, read_result,
+                           &response, &response_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the KMIP key response.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+
+  unsigned char *received_key_id = NULL;
+  size_t received_key_id_len = 0;
+
+  // Parse the key response
+  result = parse_kmip_get_response(&kmip_context,
+                                   response, response_len,
+                                   &received_key_id, &received_key_id_len,
+                                   key, key_len);
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to parse the KMIP Get response.");
+    kmip_destroy(&kmip_context);
+    return 1;
+  }
+  kmyth_log(LOG_DEBUG, "Received a KMIP object with ID: %.*s",
+            received_key_id_len, received_key_id);
+
+  kmip_destroy(&kmip_context);
+
+  return 0;
+}
+
+//
+// send_key_with_session_key()
+//
+int send_key_with_session_key(int socket_fd,
+                              unsigned char *session_key,
+                              size_t session_key_len, unsigned char *key,
+                              size_t key_len)
+{
+  unsigned char *encrypted_request = calloc(8192, sizeof(unsigned char));
```

Comment:
We should check that the return value of calloc() is not NULL

Reply:
Done